File: src/commands/list.ts
import { CAC } from 'cac'
import { resolve } from 'node:path'
import { Cirno } from '../index.ts'
import { info } from '../utils.ts'

export default (cli: CAC) => cli
  .command('list', 'List all applications')
  .alias('ls')
  .alias('tree')
  .option('--cwd <path>', 'Specify the project folder')
  .option('--json', 'Output as JSON')
  .action(async (options) => {
    const cwd = resolve(process.cwd(), options.cwd ?? '.')
    const cirno = await Cirno.init(cwd)
    const apps = cirno.data.apps
    if (options.json) return console.log(JSON.stringify(apps))
    if (!apps.length) return info('No applications found.')
    info(`Found ${apps.length} applications:`)
    apps.forEach((app, index) => {
      const prefix = index === apps.length - 1 ? '└' : '├'
      console.log(`${prefix}── ${app.id}\t${app.name}`)
      app.backups.forEach((backup, index) => {
        const prefix = index === app.backups.length - 1 ? '└' : '├'
        console.log(`    ${prefix}── ${backup.id}`)
      })
    })
  })

File: cli/src/list.rs
use anyhow::Result;
use cirno_core::Cirno;
use clap::Args;

use crate::EnvArgs;

#[derive(Debug, Args)]
pub struct List {
    #[clap(long, help = "Output in JSON format")]
    json: bool,
}

impl EnvArgs for List {
    async fn main(self, cirno: Cirno) -> Result<()> {
        if self.json {
            let json = serde_json::to_string(&cirno.manifest.apps)?;
            println!("{json}");
            return Ok(());
        }
        if cirno.manifest.apps.is_empty() {
            println!("No applications found.");
            return Ok(());
        }
        let len = cirno.manifest.apps.len();
        println!("Found {len} applications:");
        for (i, app) in cirno.manifest.apps.iter().enumerate() {
            let prefix = if i == len - 1 { "└" } else { "├" };
            println!("{}── {}\t{}", prefix, app.id, app.name);
            for (j, backup) in app.backups.iter().enumerate() {
                let prefix = if j == app.backups.len() - 1 { "└" } else { "├" };
                println!("    {}── {}", prefix, backup.id);
            }
        }
        Ok(())
    }
}
